teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,212 @@
+    - --state=free
+    - --target-state=busy
+    stdout: true
+  - name: parse-project-id
+    image: imega/jq
+    args:
+    - -r
+    - .name
+    - $(steps.boskosctl.stdoutPath)
+    stdout: true
+  - name: copy-result
+    image: alpine
+    args:
+    - cp
+    - $(steps.parse-project-id.stdoutPath)
+    - $(results.project-id.path)
+```
+
+#### Risks and Mitigations
+
+* Task authors cannot configure the redirection path, it would take extra steps to implement the following use cases: 1) write stdout of a step into `.docker/config.json` and consume the generated auth configuration in a subsequent step; or 2) extract stdout into a Task result. In both cases, if the step image provides a shell, users can specify the `Script` field to overwrite the image entrypoint to invoke the original entrypoint command and redirect the stdout into the target file. There are two main drawbacks: 1) information encapsulated by the image (e.g., entrypoint) will be "leaked" into Task specification, meaning that users cannot treat a third-party utility image as a black box and just pass in appropriate arguments, and thus creating a tight coupling between the Task and the image; 2) users will lose the ability to see stdout through Pod log API, unless they maintain a forked image to package a `tee` program.
+
+* If multiple steps output large data and there is a disk limit, users cannot reuse the disk space to store redirected output data.
+
+## Design Details
+
+The following flags will be added to the `entrypoint` command to support I/O redirection of the sub-process:
+
+* `-stdout_path`: If specified, the stdout of the sub-process will be duplicated to the given path on the local filesystem.
+
+* `-stderr_path`: If specified, the stderr of the sub-process will be duplicated to the given path on the local filesystem. It can be set to the same value as `{{stdout_path}}` so both streams are copied to the same file. However, there is no ordering guarantee on data copied from both streams.
+
+A proof-of-concept implementation is presented in [tektoncd/pipeline#3103](https://github.com/tektoncd/pipeline/pull/3103).
+
+## Alternatives
+
+* Parsing stdout/stderr into a structured format ([example](https://github.com/tektoncd/pipeline/issues/2925#issue-654319361)): This approach requires the step image to produce JSON output, which limits what images can be used. It also hides the parsing magic in Tekton, which can be hard to debug if the output is malformed.
+
+* [Allowing subsequent steps to specify a filter expression to apply to step outputs](https://github.com/tektoncd/pipeline/issues/2925#issuecomment-657529820): If there are multiple subsequent "consumer" steps, then either all consumers must use the same filter to save disk space. Also the magic of filtering will be hidden by Tekton from users, creating unnecessary complexity. It is not hard to use `Script` or add an extra step to perform filtering to achieve the same result with more transparency.
```

Comment:
another alternative would be to do nothing in particular in tekton and leave this up to the Task and image authors, which effectively means if you want to communicate data between steps, or you want to capture data for results, you need a shell or you need your binaries to support outputting to a file